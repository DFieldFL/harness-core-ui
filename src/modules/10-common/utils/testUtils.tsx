import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uikit'
import qs from 'qs'

import strings from 'strings/strings.en.yaml'
import { AppStoreProvider, useAppStoreWriter } from 'framework/hooks/useAppStore'
import type { AppStore } from 'framework/types/AppStore'
import { AUTH_ROUTE_PATH_PREFIX, StringsContext } from 'framework/exports'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): Element | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): Element | null => document.querySelector('.bp3-popover-content')

interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, any>
  defaultAppStoreValues?: Partial<AppStore>
}

export const prependAccountPath = (path: string): string => AUTH_ROUTE_PATH_PREFIX + path

const AppStoreInitializer: React.FC<{ defaultAppStoreValues: TestWrapperProps['defaultAppStoreValues'] }> = ({
  children,
  defaultAppStoreValues
}) => {
  const updateAppStore = useAppStoreWriter()
  updateAppStore(defaultAppStoreValues || {})
  return <> {children} </>
}

export const NotFound = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  const { path = '/', pathParams = {}, defaultAppStoreValues, queryParams = {} } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  const history = createMemoryHistory({ initialEntries: [routePath] })

  return (
    <AppStoreProvider>
      <StringsContext.Provider value={strings}>
        <AppStoreInitializer defaultAppStoreValues={defaultAppStoreValues}>
          <Router history={history}>
            <ModalProvider>
              <RestfulProvider base="/">
                <Switch>
                  <Route exact path={path}>
                    {props.children}
                  </Route>
                  <Route>
                    <NotFound />
                  </Route>
                </Switch>
              </RestfulProvider>
            </ModalProvider>
          </Router>
        </AppStoreInitializer>
      </StringsContext.Provider>
    </AppStoreProvider>
  )
}
