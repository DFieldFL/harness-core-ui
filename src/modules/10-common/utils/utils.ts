import type { IconName } from '@wings-software/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}
