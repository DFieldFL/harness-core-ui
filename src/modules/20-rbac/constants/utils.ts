export enum RbacResourceGroupTypes {
  DYNAMIC_RESOURCE_SELECTOR = 'DynamicResourceSelector'
}
