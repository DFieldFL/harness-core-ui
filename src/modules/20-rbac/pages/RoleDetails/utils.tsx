import produce from 'immer'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@common/interfaces/ResourceInterface'
import type { Permission, PermissionResponse } from 'services/rbac'

export const getPermissionMap = (permissionList?: PermissionResponse[]): Map<ResourceType, Permission[]> => {
  const permissionsMap: Map<ResourceType, Permission[]> = new Map()
  permissionList?.map(response => {
    if (response.permission.resourceType) {
      if (permissionsMap.has(response.permission.resourceType as ResourceType)) {
        permissionsMap.get(response.permission.resourceType as ResourceType)?.push(response.permission)
      } else permissionsMap.set(response.permission.resourceType as ResourceType, [response.permission])
    }
  })
  return permissionsMap
}

export const onPermissionChange = (
  permission: string,
  isAdd: boolean,
  permissions: string[],
  setPermissions: (value: React.SetStateAction<string[]>) => void
): void => {
  if (isAdd) {
    if (
      permission === PermissionIdentifier.EDIT_DASHBOARD ||
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    ) {
      setPermissions(_permissions => [...permissions, permission, PermissionIdentifier.VIEW_DASHBOARD])
    } else {
      setPermissions(_permissions => [...permissions, permission])
    }
  } else if (
    !(
      permission === PermissionIdentifier.VIEW_DASHBOARD &&
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    )
  ) {
    setPermissions(_permissions =>
      produce(_permissions, draft => {
        draft?.splice(permissions.indexOf(permission), 1)
      })
    )
  }
}
