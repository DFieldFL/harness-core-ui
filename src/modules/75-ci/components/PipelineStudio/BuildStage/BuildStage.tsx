import React from 'react'
import { PipelineStage } from '@pipeline/components/PipelineStages/PipelineStage'
import type { BuildStageElementConfig, StageElementWrapper } from '@pipeline/utils/pipelineTypes'
import { EditStageView } from './EditStageView/EditStageView'
import BuildStageSetupShell from '../BuildStageSetupShell/BuildStageSetupShell'

interface BuildStageProps {
  data?: StageElementWrapper<BuildStageElementConfig>
  onSubmit?: (values: StageElementWrapper<BuildStageElementConfig>, identifier: string) => void
}

export class BuildStage extends PipelineStage<BuildStageProps> {
  render(): JSX.Element {
    const { minimal, stageProps } = this.props
    if (minimal) {
      return <EditStageView {...stageProps} />
    }
    return <BuildStageSetupShell />
  }
}
