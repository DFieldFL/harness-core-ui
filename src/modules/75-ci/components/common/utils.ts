import { parse } from 'yaml'
import { get } from 'immer/dist/internal'
import { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
import { loggerFor } from 'framework/logging/logging'
import type { CompletionItemInterface } from '@common/interfaces/YAMLBuilderProps'
import { ModuleName } from 'framework/types/ModuleName'
import { getConnectorSuggestions } from '../PipelineSteps/EditorSuggestionUtils'

const logger = loggerFor(ModuleName.CI)

/* istanbul ignore next */
export const getConnectorList = async (
  connectorTypes: string[] | undefined,
  path: string,
  yaml: string,
  params: Record<string, unknown>
): Promise<CompletionItemInterface[]> => {
  if (connectorTypes?.length) {
    return getConnectorSuggestions(params, connectorTypes)
  }
  let pipelineObj
  try {
    pipelineObj = parse(yaml)
  } catch (err) {
    logger.error('Error while parsing the yaml', err)
  }
  if (pipelineObj) {
    const obj = get(pipelineObj, path.replace('.spec.connectorRef', ''))
    if (obj.type === StepType.Dependency) {
      return getConnectorSuggestions(params, ['Gcp', 'Aws', 'DockerRegistry'])
    }
  }
  return []
}
