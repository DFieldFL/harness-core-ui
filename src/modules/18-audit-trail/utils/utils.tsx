import React from 'react'
import { Menu } from '@blueprintjs/core'
import type { ItemRenderer } from '@blueprintjs/select'
import { Layout, Text, Avatar, Color, MultiSelectOption } from '@wings-software/uicore'
import { isEmail } from '@common/utils/Validation'

export interface UserItem extends MultiSelectOption {
  email?: string
}

export const UserItemRenderer: ItemRenderer<UserItem> = (item, { handleClick }) => (
  <Menu.Item
    key={item.value.toString()}
    text={
      <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Avatar name={item.label} email={item.email || item.value.toString()} size="small" hoverCard={false} />
        <Layout.Vertical padding={{ left: 'small' }}>
          <Text color={Color.BLACK}>{item.label}</Text>
          <Text color={Color.GREY_700}>{item.email || item.value}</Text>
        </Layout.Vertical>
      </Layout.Horizontal>
    }
    onClick={handleClick}
  />
)

export const UserTagRenderer = (item: UserItem, validate = false): React.ReactNode => (
  <Layout.Horizontal key={item.value.toString()} flex spacing="small">
    <Avatar name={item.label} email={item.value.toString()} size="xsmall" hoverCard={false} />
    <Text color={validate && !isEmail(item.value.toString().toLowerCase()) ? Color.RED_500 : Color.BLACK}>
      {item.label}
    </Text>
  </Layout.Horizontal>
)
