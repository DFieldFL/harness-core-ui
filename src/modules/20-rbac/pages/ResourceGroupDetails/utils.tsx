import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { ResourceSelector } from 'services/cd-ng'

export const getSelectedResourcesMap = (
  resourceSelectorList?: ResourceSelector[]
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()

  resourceSelectorList?.map(resourceSelector => {
    if (resourceSelector.type === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      map.set(resourceSelector.resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    else map.set(resourceSelector.resourceType, resourceSelector.identifiers)
  })

  return map
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelector[] => {
  const resourceSelectors: ResourceSelector[] = []
  map.forEach((value, key) => {
    if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      resourceSelectors.push({
        type: RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR,
        resourceType: key
      })
    else {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.STATIC_RESOURCE_SELECTOR,
        resourceType: key,
        identifiers: value
      })
    }
  })
  return resourceSelectors
}
