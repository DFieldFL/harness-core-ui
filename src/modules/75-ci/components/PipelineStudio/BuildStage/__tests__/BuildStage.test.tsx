import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { BuildStage } from '../BuildStage'

jest.mock('../EditStageView/EditStageView', () => ({ EditStageView: () => <div /> }))
jest.mock('../../BuildStageSetupShell/BuildStageSetupShell', () => () => <div />)
jest.mock('@common/components/YAMLBuilder/YamlBuilder')
jest.mock('@common/utils/YamlUtils', () => ({}))

describe('BuildStage', () => {
  test('matches snapshot', () => {
    const { container } = render(
      <TestWrapper>
        <BuildStage
          name="test"
          type="ADVANCED"
          icon="ci-main"
          isDisabled={false}
          isApproval={false}
          title="test"
          description=""
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('matches snapshot when minimal="true"', () => {
    const { container } = render(
      <TestWrapper>
        <BuildStage
          name="test"
          type="ADVANCED"
          icon="ci-main"
          isDisabled={false}
          isApproval={false}
          title="test"
          description=""
          minimal
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
