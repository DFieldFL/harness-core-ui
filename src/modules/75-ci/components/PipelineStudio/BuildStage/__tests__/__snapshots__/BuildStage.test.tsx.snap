// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildStage matches snapshot 1`] = `
<div>
  <div />
</div>
`;

exports[`BuildStage matches snapshot when minimal="true" 1`] = `
<div>
  <div />
</div>
`;
